browser/ui/webui/brave_new_tab/new_tab_page_handler.cc

```diff
@@ -0,0 +1,421 @@
+#include "brave/components/ntp_background_images/browser/url_constants.h"
+#include "brave/components/ntp_background_images/browser/view_counter_service.h"
+#include "brave/components/ntp_background_images/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+
+namespace brave_new_tab {
+
+namespace {
+
+std::string GetCustomImageURL(const std::string& image_name) {
+  return CustomBackgroundFileManager::Converter(image_name).To<GURL>().spec();
+}
+
+std::string CustomImageNameFromURL(const std::string& url) {
+  return CustomBackgroundFileManager::Converter(GURL(url)).To<std::string>();
+}
+
+mojom::SponsoredImageBackgroundPtr ReadSponsoredImageData(
+    const base::Value::Dict& data) {
+  using ntp_background_images::kAltKey;
+  using ntp_background_images::kCampaignIdKey;
+  using ntp_background_images::kCreativeInstanceIDKey;
+  using ntp_background_images::kDestinationURLKey;
+  using ntp_background_images::kImageKey;
+  using ntp_background_images::kIsBackgroundKey;
+  using ntp_background_images::kLogoKey;
+  using ntp_background_images::kWallpaperIDKey;
+  using ntp_background_images::kWallpaperImageURLKey;
+
+  auto is_background = data.FindBool(kIsBackgroundKey);
+  if (is_background.value_or(false)) {
+    return nullptr;
+  }
+
+  auto background = mojom::SponsoredImageBackground::New();
+
+  if (auto* creative_instance_id = data.FindString(kCreativeInstanceIDKey)) {
+    background->creative_instance_id = *creative_instance_id;
+  }
+
+  if (auto* wallpaper_id = data.FindString(kWallpaperIDKey)) {
+    background->wallpaper_id = *wallpaper_id;
+  }
+
+  if (auto* campaign_id = data.FindString(kCampaignIdKey)) {
+    background->campaign_id = *campaign_id;
+  }
+
+  if (auto* image_url = data.FindString(kWallpaperImageURLKey)) {
+    background->image_url = *image_url;
+  }
+
+  if (auto* logo_dict = data.FindDict(kLogoKey)) {
+    auto logo = mojom::SponsoredImageLogo::New();
+    if (auto* alt = logo_dict->FindString(kAltKey)) {
+      logo->alt = *alt;
+    }
+    if (auto* destination_url = logo_dict->FindString(kDestinationURLKey)) {
+      logo->destination_url = *destination_url;
+    }
+    if (auto* image_url = logo_dict->FindString(kImageKey)) {
+      logo->image_url = *image_url;
+    }
+    if (!logo->image_url.empty()) {
+      background->logo = std::move(logo);
+    }
+  }
+
+  return background;
+}
+
+}  // namespace
+
+NewTabPageHandler::NewTabPageHandler(
+    mojo::PendingReceiver<mojom::NewTabPageHandler> receiver,
+    std::unique_ptr<CustomImageChooser> custom_image_chooser,
+    std::unique_ptr<CustomBackgroundFileManager> custom_file_manager,
+    PrefService* pref_service,
+    ntp_background_images::ViewCounterService* view_counter_service,
+    bool is_restored_page)
+    : receiver_(this, std::move(receiver)),
+      update_observer_(pref_service,
+                       base::BindRepeating(&NewTabPageHandler::OnUpdate,
+                                           base::Unretained(this))),
```

Comment:
these instances seem alright at the moment, but it's only because of the way we're setting up the pipe to create a new instance on the JS side I believe: https://github.com/brave/brave-core/pull/26408/files/6ad63cc7cbe9685620a88f93953b6c2c71f36735#diff-55b514ae43245d278d5b48c5740573e3c8918ee65439c790bcd87daa9e8f2cf9R20-R28

If this changes, I think it produces a UAF issue, so the safer option seems to be to switch this to use a weak_ptr so that we can avoid an issue if that JS code is refactored based on what I'm seeing.

Reply:
Currently, all of these `Unretained`-using callbacks are passed to objects that the `NewTabPageHandler` owns, and they aren't supposed to allow callback execution after destruction. However, it's a little hard to see this and verify, so I think it would be a good idea to use weak pointers here.

Reply:
Fixed